### PR TITLE
[BACK-46] Create dbUpdater

### PR DESCRIPTION
**Pull request by thomas-woehrle**
### Description of Issue
Repositories stored in the database should be updated on a daily basis. 

Depending on their status repos in the db have to be deleted or updated. New trending repos have to be added to the db. 

This function should integrate the other functionalities (like getting eli5, github data etc.)

**This PR also entails big refactoring.** 
### How I implemented
- generated supabase types see https://supabase.com/docs/guides/api/rest/generating-types
- the dbUpdater function goes through several steps. 
1. the trending repos are fetched 
2. for each repo it is checked whether it is already in the db
3. if not data is aggregated for it and its owner
4. Both are added to the db

**Refactoring:**
- new api folder for functions which only access apis 
- new scraping folder for functions which only do scraping
- refactoring of existing functions into these folders
- tests folder for tests
- types folder and refactoring types of existing types into their respective files
- I refactored everything that has to do with starHistory in a new folder because I don't want to break it considering it comes pretty similar (identical) to star-history's version
### Other 
First & Second version of the dbUpdater come in this RR.
Following versions will:
- incorporate additional data
- add users as owners if applicable
- get data for weekly and monthly trending repos

### CODE REVIEW

**Comment by pczern** [comment]
please remove, it will interfere with the new GitHub Action, that will check for good PR formatting
> will fill out once development proceeds.
> will fill out once development proceeds
> will fill out once development proceeds
> will fill out once development proceeds

File: packages/graphql_server/src/dbUpdater.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  const { data: existingRepos, error: error2 } = await supabase
+    .from('project')
+    .select('name, owned_by ( name )')
+
+  // update the github statistics for all projects that remain in the database
+  if (existingRepos) {
+    updateRepos(existingRepos.map((repo) => ({ name: repo.name, owner: repo.owned_by?.name })))
+  }
+
+  // get trending repos from github
```
Comment (pczern) [inline packages/graphql_server/src/dbUpdater.ts:30]: add `// @TODO` for such things

File: packages/graphql_server/src/dbUpdater.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  cutoffTime.setMinutes(cutoffTime.getMinutes() - minutes)
+  return cutoffTime.toISOString()
+}
+
+const updateRepos = (repos: Repository[]) => {
+  return null
+}
+
+const getTrendingRepos = () => {
+  return
```
Comment (pczern) [inline packages/graphql_server/src/dbUpdater.ts:52]: Remove `return`, remove function, add `// @Todo`

File: packages/graphql_server/types/supabase.ts
Comment (pczern) [inline packages/graphql_server/types/supabase.ts:1]: well done

File: packages/graphql_server/src/dbUpdater.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+dbUpdater()
```
Comment (pczern) [inline packages/graphql_server/src/dbUpdater.ts:55]: add `void` before call

File: packages/graphql_server/package.json
Comment (pczern) [inline packages/graphql_server/package.json:1]: add command to generate supabase typescript types

File: packages/graphql_server/src/dbUpdater.ts
Comment (pczern) [inline packages/graphql_server/src/dbUpdater.ts:1]: this file is not finished

**Review by pczern — CHANGES_REQUESTED** [review]

**Comment by pczern** [comment]
@thomas-woehrle address review, update description

File: packages/graphql_server/package.json
Comment (thomas-woehrle) [reply packages/graphql_server/package.json:1]: Will do it don't have time for it today though sorry. 
Could you please still review the new changes

**Comment by thomas-woehrle** [comment]
It is ready to be reviewed although one thing is missing still.

File: packages/graphql_server/package.json
Comment (thomas-woehrle) [reply packages/graphql_server/package.json:1]: Could I maybe add it in the next version of dbUpdater, which i will start implementing tomorrow. I have really no more time today and it would be important to get this PR merged soon so everyone can integrate it into their branches
.

File: packages/graphql_server/src/dbUpdater.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+) => {
+  const cutoffTime = new Date()
+  cutoffTime.setHours(cutoffTime.getHours() - hours)
+  cutoffTime.setMinutes(cutoffTime.getMinutes() - minutes)
+  return cutoffTime.toISOString()
+}
+
+// daily: gh + twitter
+// weekly: rest
+// @TODO: imlement
```
Comment (cnk1dev) [inline packages/graphql_server/src/dbUpdater.ts:61]:
```suggestion
// @TODO: implement
```

File: packages/graphql_server/src/dbUpdater.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+const updateRepo = (name: string, owner: string) => {
+  return null
+}
+
+/**
+ * Goes throuhg the list of repos and processes them one by one.
```
Comment (cnk1dev) [inline packages/graphql_server/src/dbUpdater.ts:67]:
```suggestion
 * Goes through the list of repos and processes them one by one.
```

File: packages/graphql_server/src/supabase.ts [hunk 6]
```diff
@@ -0,0 +1,8 @@
+import { createClient } from '@supabase/supabase-js'
+import { Database } from '../types/supabase'
+
+const supabase = createClient<Database>(process.env.SUPABASE_URL, process.env.SUPABASE_API_KEY)
+
+// const supabase = createClient<Database>('url', 'key')
```
Comment (cnk1dev) [inline packages/graphql_server/src/supabase.ts:6]:
```suggestion
```

**Review by cnk1dev — APPROVED** [review]

**Comment by cnk1dev** [comment]
Details like folder structure can be discussed & adapted later but PR should be merged asap to test & have demo ready for midterm review

File: packages/graphql_server/src/api/githubApi.ts
Comment (pczern) [inline packages/graphql_server/src/api/githubApi.ts:1]: don't catch errors here, let the requester of data catch them

File: packages/graphql_server/src/dataAggregation.ts [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    name: repoGHdata.name,
+    about: repoGHdata.description,
+    star_count: repoGHdata.stargazerCount,
+    issue_count: repoGHdata.issues.totalCount,
+    fork_count: repoGHdata.forkCount,
+    pull_request_count: repoGHdata.pullRequests.totalCount,
+    contributor_count: 1,
+    github_url: repoGHdata.url,
+    website_url: repoGHdata.homepageUrl,
+    owned_by: organizationID ?? '634b6eb5-30c8-4818-81d3-e1d98cb0b2c7',
```
Comment (pczern) [inline packages/graphql_server/src/dataAggregation.ts:57]: no

File: packages/graphql_server/src/dataAggregation.ts [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+  // @Todo aggregate more data for the repo
+
+  repoInfo = {
+    name: repoGHdata.name,
+    about: repoGHdata.description,
+    star_count: repoGHdata.stargazerCount,
+    issue_count: repoGHdata.issues.totalCount,
+    fork_count: repoGHdata.forkCount,
+    pull_request_count: repoGHdata.pullRequests.totalCount,
```
Comment (pczern) [inline packages/graphql_server/src/dataAggregation.ts:53]: use repoGHdata?.pullRequests?.totalCount for every assignable, what if the api changes and our app crashes, because of one datapoint

File: packages/graphql_server/src/dataAggregation.ts [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+/**
+ * Returns the organization id for the given organization name.
+ * It tries to get that id from the DB. If it is not in the DB it gets the data from github and inserts it into the DB.
+ * @param {string} owner - The name of the organization.
+ * @returns {string} The id of the organization.
+ */
+const getOrganizationID = async (owner: string) => {
+  const { data: organization, error: getOrganizationError } = await supabase
```
Comment (pczern) [inline packages/graphql_server/src/dataAggregation.ts:71]: the naming guidelines state that methods should be named with verbs, this is a property and should be named `organizationError` if at all or just `error`, `supabaseError` or if at all `selectOrganizationError` so it's understandable

File: packages/graphql_server/src/dataAggregation.ts [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    .from('organization')
+    .select('id')
+    .eq('name', owner)
+  getOrganizationError &&
+    console.error('Error getting organization', owner, 'from database: \n', getOrganizationError)
+
+  // if a organization with this name is already in the database return the id
+  if (organization && organization[0]) {
```
Comment (pczern) [inline packages/graphql_server/src/dataAggregation.ts:79]: one line expression
`if(organization?.[0]?.id) return organization[0].id`

File: packages/graphql_server/src/dataAggregation.ts [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+
+    const { error: insertOrganizationError } = await supabase
+      .from('organization')
+      .insert([organizationDataDBFormat])
+    insertOrganizationError &&
+      console.error('Error inserting organization into database: \n', insertOrganizationError)
+
+    const { data: orga } = await supabase.from('organization').select('id').eq('name', owner)
+
+    if (orga && orga[0]) {
```
Comment (pczern) [inline packages/graphql_server/src/dataAggregation.ts:122]: if (orga?.[0]?.id) return orga?.[0]?.id; what is being returned in every other case? Shouldn't this throw an error then?

File: packages/graphql_server/src/processRepo.ts [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+      .eq('id', repo.owned_by)
+
+    // the owner has the same name -> the repo is already in the database
+    // atm also checks whether the owner is vercel because vercel is inserted whenever the owner is a user not a organization
+    if (
+      currentOrg &&
+      currentOrg[0] &&
+      currentOrg[0].login !== 'vercel' &&
+      currentOrg[0].login === owner
+    ) {
```
Comment (pczern) [inline packages/graphql_server/src/processRepo.ts:73]: if(currentOrg?.[0]?.login !== 'vercel' && currentOrg?.[0]?.login === owner)
    return true

**Review by pczern — CHANGES_REQUESTED** [review]
rename all types to `d.ts`

File: packages/graphql_server/src/dataAggregation.ts (hunk 7 above)
Comment (thomas-woehrle) [reply packages/graphql_server/src/dataAggregation.ts:57]: this is just a übergangslösung. Right now when no organization can be found as an owner for example because the owner is a user, this id which belongs to vercel in the db is assigned as owner. 
I will change this today but would prefer to do so on a new branch where i also add other features. 
This PR should be merged asap

File: packages/graphql_server/src/dataAggregation.ts (hunk 11 above)
Comment (thomas-woehrle) [reply packages/graphql_server/src/dataAggregation.ts:122]: Good point, I decided to return null in this case. Actually this case should never happen. Because if the organization does not exist it will be inserted.

**Comment by thomas-woehrle** [comment]
Philipp what is wrong with the current approach of naming the types files?

**Comment by thomas-woehrle** [comment]
Could you please review the new version asap even if the github action for updating supabse types does not exist yet? Because this PR introduces changes to the folder/file structure that are importannt for everyone

File: packages/graphql_server/src/dataAggregation.ts [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+
+    const { error: organizationInsertionError } = await supabase
+      .from('organization')
+      .insert([organizationDataDBFormat])
+    organizationInsertionError &&
+      console.error('Error inserting organization into database: \n', organizationInsertionError)
+
+    const { data: orga } = await supabase.from('organization').select('id').eq('login', owner)
+
+    if (orga?.[0]?.id) {
```
Comment (pczern) [inline packages/graphql_server/src/dataAggregation.ts:125]: ternary

**Review by pczern — APPROVED** [review]
